### PR TITLE
Type definitions for JSZip

### PR DESCRIPTION
**Pull request by deliangyang**
Suport typescript Type definitions

### CODE REVIEW

**Comment by deliangyang** [comment]
Sorry to ping @Stuk could this PR be reviewed?

File: index.d.ts [hunk 1]
```diff
@@ -0,0 +1,269 @@
...
+
+interface CompressedObject {
+    compressedSize: number;
+    uncompressedSize: number;
+    crc32: number;
+    compression: object;
+    compressedContent: string|ArrayBuffer|Uint8Array|Buffer;
+}
+
+type InputFileFormat = InputByType[keyof InputByType];
```
Comment (Stuk) [inline index.d.ts]: What does this mean?

File: index.d.ts [hunk 2]
```diff
@@ -0,0 +1,269 @@
...
+        dir: boolean;
+        date: Date;
+        comment: string;
+        /** The UNIX permissions of the file, if any. */
+        unixPermissions: number | string | null;
+        /** The UNIX permissions of the file, if any. */
+        dosPermissions: number | null;
+        options: JSZipObjectOptions;
+
+        _data: CompressedObject;
```
Comment (Stuk) [inline index.d.ts]: This is private, so I don't think it should be part of the TS interface.

File: index.d.ts [hunk 3]
```diff
@@ -0,0 +1,269 @@
...
+
+    /**
+     * Add a file to the archive
+     *
+     * @param path Relative path to file
+     * @param data Content of the file
+     * @param options Optional information about the file
+     * @return JSZip object
+     */
+    file<T extends JSZip.InputType>(path: string, data: InputByType[T] | Promise<InputByType[T]>, options?: JSZip.JSZipFileOptions): this;
```
Comment (Stuk) [inline index.d.ts]: On use does one need to write `file<Blob>(...)`, or will it be inferred from the `data` argument?

File: index.d.ts [hunk 4]
```diff
@@ -0,0 +1,269 @@
...
+    generateAsync<T extends JSZip.OutputType>(options?: JSZip.JSZipGeneratorOptions<T>, onUpdate?: OnUpdateCallback): Promise<OutputByType[T]>;
+
+    /**
+     * Generates a new archive asynchronously
+     *
+     * @param options Optional options for the generator
+     * @param onUpdate The optional function called on each internal update with the metadata.
+     * @return A Node.js `ReadableStream`
+     */
+    generateNodeStream(options?: JSZip.JSZipGeneratorOptions<'nodebuffer'>, onUpdate?: OnUpdateCallback): NodeJS.ReadableStream;
```
Comment (Stuk) [inline index.d.ts]: Interesting, so the type is a string (`nodebuffer`) instead of the type itself?

File: index.d.ts (hunk 2 above)
Comment (deliangyang) [reply index.d.ts]: Yes, but sometimes we need to get the compression size and origin file's size.

File: index.d.ts (hunk 1 above)
Comment (deliangyang) [reply index.d.ts]:
```suggestion
interface InputByType {
    base64: string;
    string: string;
    text: string;
    binarystring: string;
    array: number[];
    uint8array: Uint8Array;
    arraybuffer: ArrayBuffer;
    blob: Blob;
    stream: NodeJS.ReadableStream;
}
```

InputFileFormat's type is ```string | number[] | Unit8Array |  Blob | NodeJS.ReadableStream```

Comment (deliangyang) [reply index.d.ts]: InputFileFormat's type is the element of interface InputByType

File: index.d.ts (hunk 3 above)
Comment (deliangyang) [reply index.d.ts]: It'll be inferred from the data argument

File: index.d.ts (hunk 4 above)
Comment (deliangyang) [reply index.d.ts]: Yes.

File: index.d.ts (hunk 2 above)
Comment (Stuk) [reply index.d.ts]: Hmm. I'm not comfortable explicitly exposing deliberately private properties.

Could you remove this property from this PR, and open another PR that exposes this property explicitly?

File: index.d.ts (hunk 4 above)
Comment (Stuk) [reply index.d.ts]: Why is this a string and not the type?

**Comment by Regnander** [comment]
Any updates on this?

**Comment by re2005** [comment]
Hi guys, thanks for the awesome JSZip.

I'm using TS and wonder if you guys have intention to merge this PR ?
Would be great.

Thanks

**Comment by sschmidTU** [comment]
We're using TypeScript and JSZip 3.4.0 failed our CI build. Needed just a little configuration change, but still a breaking change.

before (breaking with 3.4.0):
```typescript
import JSZip = require("jszip");
// starting with jszip 3.4.0, JSZip.JSZip is not found,
//    probably because of new possibly conflicting TypeScript definitions
const zip: JSZip.JSZip = new JSZip();
// (zip.load...)
return zip.file("test.xml").async("string");
```

now working:
```typescript
import JSZip from "jszip";
const zip: JSZip = new JSZip();
// (zip.load...)
return zip.file("test.xml").async("text");
```
This may also have to do with us using
`"esModuleInterop": true,`
in tsconfig.json since recently, but still, jszip 3.3.0 was still passing builds.

I also was a bit surprised that "string" was suddenly not valid anymore, the typescript definition allows various strings now, "text" one of them.

also, using the package es6-promise caused a conflict with Promise, but it was apparently unnecessary anyways.

**Pull request merged**